src/gen/conditions.rs

```diff
@@ -0,0 +1,1568 @@
+            a,
+            first(a, c)?,
+            ErrorCode::AssertHeightRelative,
+            range_cache,
+            flags,
+        )?)),
+        ASSERT_HEIGHT_ABSOLUTE => Ok(Condition::AssertHeightAbsolute(parse_height(
+            a,
+            first(a, c)?,
+            ErrorCode::AssertHeightAbsolute,
+            range_cache,
+            flags,
+        )?)),
+        _ => Err(ValidationErr(c, ErrorCode::InvalidConditionOpcode)),
+    }
+}
+
+#[derive(Debug)]
+pub struct Npc {
+    pub coin_id: Arc<[u8; 32]>,
+    pub puzzle_hash: NodePtr,
+    // conditions
+    // The sum of all reserve fee conditions
+    pub reserve_fee: u64,
+    // the highest height/time conditions (i.e. most strict)
+    pub height_relative: u32,
+    pub height_absolute: u32,
+    pub seconds_relative: u64,
+    pub seconds_absolute: u64,
+    // all create coins. Duplicates are consensus failures
+    pub create_coin: HashSet<(Vec<u8>, u64)>,
+    // Agg Sig conditions
+    pub agg_sigs: HashSet<Condition>,
+}
+
+struct AnnounceState {
+    // hashing of the announcements is deferred until parsing is complete. This
+    // means less work up-front, in case parsing/validation fails
+    coin: HashSet<(Arc<[u8; 32]>, NodePtr)>,
+    puzzle: HashSet<(NodePtr, NodePtr)>,
+
+    // the assert announcements are checket once everything has been parsed and
+    // validated.
+    assert_coin: HashSet<NodePtr>,
+    assert_puzzle: HashSet<NodePtr>,
+}
+
+impl AnnounceState {
+    fn new() -> AnnounceState {
+        AnnounceState {
+            coin: HashSet::new(),
+            puzzle: HashSet::new(),
+            assert_coin: HashSet::new(),
+            assert_puzzle: HashSet::new(),
+        }
+    }
+}
+
+fn parse_npc(
+    a: &Allocator,
+    ann: &mut AnnounceState,
+    spent_coins: &mut HashSet<Arc<[u8; 32]>>,
+    mut spend: NodePtr,
+    flags: u32,
+    max_cost: &mut Cost,
+    range_cache: &mut RangeSet,
+) -> Result<Npc, ValidationErr> {
+    let parent_id = sanitize_hash(a, first(a, spend)?, 32, ErrorCode::InvalidParentId)?;
+    spend = rest(a, spend)?;
+    let puzzle_hash = sanitize_hash(a, first(a, spend)?, 32, ErrorCode::InvalidPuzzleHash)?;
+    spend = rest(a, spend)?;
+    let amount_buf = sanitize_integer(
+        a,
+        first(a, spend)?,
+        8,
+        ErrorCode::InvalidCoinAmount,
+        range_cache,
+        flags,
+    )?
+    .to_vec();
+    let my_amount = u64_from_bytes(&amount_buf);
+    let cond = rest(a, spend)?;
+    let coin_id = Arc::new(compute_coin_id(a, parent_id, puzzle_hash, &amount_buf));
+
+    if !spent_coins.insert(coin_id.clone()) {
+        // if this coin ID has already been added to this set, it's a double
+        // spend
+        return Err(ValidationErr(spend, ErrorCode::DoubleSpend));
+    }
+
+    let mut npc = Npc {
+        coin_id,
+        puzzle_hash,
+        reserve_fee: 0,
+        height_relative: 0,
+        height_absolute: 0,
+        seconds_relative: 0,
```

Comment:
Can `(ASSERT_RELATIVE_SECONDS 0)` ever fail? If block timestamps are not purely monotonic, it could fail, which would make this a soft fork (ie. `(ASSERT_RELATIVE_SECONDS 0)` is different from no `ASSERT_RELATIVE_SECONDS` of any kind.

I know in bitcoin block timestamps are not purely monotonically increasing. @mariano54

Reply:
right, my assumption is that all `ASSERT_`(`SECONDS`|`HEIGHT`)`_`(`ABSOLUTE`|`RELATIVE`) are inherently true if the argument is <= 0

Reply:
this has been updated for `height_relative`. Since 0 *is* relevant, I use an `Option<u32>` there now